jadx-core/src/main/java/jadx/core/utils/kotlin/KotlinMetadataConsts.kt

```diff
@@ -0,0 +1,83 @@
+@file:Suppress("UNCHECKED_CAST")
+package jadx.core.utils.kotlin
+
+import jadx.api.plugins.input.data.annotations.EncodedType
+import jadx.api.plugins.input.data.annotations.EncodedValue
+import jadx.api.plugins.input.data.annotations.IAnnotation
+import jadx.core.dex.nodes.ClassNode
+import kotlinx.metadata.jvm.KotlinClassMetadata
+import kotlinx.metadata.jvm.Metadata
+
+
+object KotlinMetadataConsts {
+	const val KOTLIN_METADATA_ANNOTATION = "Lkotlin/Metadata;"
+	const val KOTLIN_METADATA_K_PARAMETER = "k"
+	const val KOTLIN_METADATA_D1_PARAMETER = "d1"
+	const val KOTLIN_METADATA_D2_PARAMETER = "d2"
+	const val KOTLIN_METADATA_MV_PARAMETER = "mv"
+	const val KOTLIN_METADATA_XS_PARAMETER = "xs"
+	const val KOTLIN_METADATA_PN_PARAMETER = "pn"
+	const val KOTLIN_METADATA_XI_PARAMETER = "xi"
+}
+
+fun ClassNode.getMetadata(): Metadata? {
+	val annotation: IAnnotation? = getAnnotation(KotlinMetadataConsts.KOTLIN_METADATA_ANNOTATION)
+
+	return annotation?.run {
+		val k = getParamAsInt(KotlinMetadataConsts.KOTLIN_METADATA_K_PARAMETER)
+		val mvArray = getParamAsIntArray(KotlinMetadataConsts.KOTLIN_METADATA_MV_PARAMETER)
+		val d1Array = getParamAsStringArray(KotlinMetadataConsts.KOTLIN_METADATA_D1_PARAMETER)
+		val d2Array = getParamAsStringArray(KotlinMetadataConsts.KOTLIN_METADATA_D2_PARAMETER)
+		val xs = getParamAsString(KotlinMetadataConsts.KOTLIN_METADATA_XS_PARAMETER)
+		val pn = getParamAsString(KotlinMetadataConsts.KOTLIN_METADATA_PN_PARAMETER)
+		val xi = getParamAsInt(KotlinMetadataConsts.KOTLIN_METADATA_XI_PARAMETER)
+
+		Metadata(
+			kind = k,
+			metadataVersion = mvArray,
+			data1 = d1Array,
+			data2 = d2Array,
+			extraString = xs,
+			packageName = pn,
+			extraInt = xi
+		)
+	}
+}
+
+private fun IAnnotation.getParamsAsList(paramName: String): List<EncodedValue>? {
+	val encodedValue = values[paramName]
+		?.takeIf { it.type == EncodedType.ENCODED_ARRAY && it.value is List<*> }
+	return encodedValue?.value?.let { it as List<EncodedValue> }
+}
+
+private fun IAnnotation.getParamAsStringArray(paramName: String): Array<String>? {
+	return getParamsAsList(paramName)
+		?.map<EncodedValue, Any?>(EncodedValue::getValue)
+		?.onEach { if (it != null && it !is String) /* TODO is this valid ? */ return@onEach }
+		?.map { "$it" }
+		?.toTypedArray()
```

Comment:
Well, if `EncodedValue::getType` return `ENCODED_STRING` so returned value should be a `String`.
And you don't need to create a string just use cast.

Reply:
The d2 field in kotlin.Metadata is Array<String> type, so I expected every element in it to be of String type, but the EncodedValue parser sometimes converted it to other types.

Can't filter those as array size needs to stay the same, otherwise i get `ArrayIndexOutOfBoundsException` from the `toKmClass()` method. I'm not sure if thats the correct way to convert that EncodedValue to the original string it was decoded from, i did not notice anything being broken because of it.